### PR TITLE
[SCM-925] Implement RemoveCommand in maven-scm-provider-jgit

### PR DESCRIPTION
**Pull request by G-Ork**
Fourth try to fix [SCM-925]

- Implement remove command for JGit provider
- Write TCK for remove command.
- No other upgrades (Java 1.8 or JGit)

### CODE REVIEW

**Comment by michael-o** [comment]
Looks much better now, will happily review.

Did you try, does this apply cleanly to 1.x? If yes, I'd back port and deliver with upcoming 1.13.0.

**Comment by G-Ork** [comment]
I did not try any back ports. Where do i find the code base for 1.x? The tag i tried are broken so i lost patience. 
There could only be two possible problems in back porting:

1. No Java 1.8 (NIO) Support. I used NIO in test because org.codehaus.plexus.util.FileUtils.fileAppend(String, String) was deprecated. Could easyly be changed.
2.  JGIT does not support remove in the older verison.

**Comment by michael-o** [comment]
> I did not try any back ports. Where do i find the code base for 1.x? The tag i tried are broken so i lost patience. There could only be two possible problems in back porting:
> 
>     1. No Java 1.8 (NIO) Support. I used NIO in test because org.codehaus.plexus.util.FileUtils.fileAppend(String, String) was deprecated. Could easyly be changed.

Look into the `maven-scm-1.x` branch. It is still Java 7. If you want your change with Maven SCM 1.13.0 delivered then use Plexus Utils, the deprecation is OK for now. We will likely update all in one go.

>     2. JGIT does not support remove in the older verison.

The currently used JGit version does not support what you need?

**Comment by G-Ork** [comment]
I just enumerate the theoretical edges. I did not verify it. For myself i do not see any advantages to back port this to an very old maven. But you may have better understanding how many may use Maven 1 and  JGit. If you could give good reason to port i will help.

**Comment by michael-o** [comment]
Maven 1? I was talking about Maven SCM 1.x.

**Comment by michael-o** [comment]
@G-Ork I am talking about https://github.com/apache/maven-scm/tree/maven-scm-1.x.

**Comment by G-Ork** [comment]
Yes i understand (now) what revision you address. I first have to setup an environment to be able to compile this branch. Is there a good reason not using the toolchain plugin ? It feel like Indiana Jones to dig out an java 1.4 JDK for checking correct back port. 

Is the back-port mandatory for this PR? 
Can we separate this in another PR?

I will free time time at Wednesday evening.

**Comment by michael-o** [comment]
> Yes i understand (now) what revision you address. I first have to setup an environment to be able to compile this branch. Is there a good reason not using the toolchain plugin ? It feel like Indiana Jones to dig out an java 1.4 JDK for checking correct back port.

No magic required, just switch and build. It will work.

> Is the back-port mandatory for this PR?

No, I will leave this decision totally up to you.

> Can we separate this in another PR?

Yes, if you decide to back port.

File: maven-scm-providers/maven-scm-providers-git/maven-scm-provider-jgit/src/main/java/org/apache/maven/scm/provider/git/jgit/command/remove/JGitRemoveCommand.java [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+        {
+            git = JGitUtils.openRepo( fileSet.getBasedir() );
+
+            List<ScmFile> removedFiles = JGitUtils.removeAllFiles( git, fileSet );
+
+            if ( logger.isDebugEnabled() )
+            {
+                for ( ScmFile scmFile : removedFiles )
+                {
+                    logger.info( "removed file: " + scmFile );
```
Comment (michael-o) [inline maven-scm-providers/maven-scm-providers-git/maven-scm-provider-jgit/src/main/java/org/apache/maven/scm/provider/git/jgit/command/remove/JGitRemoveCommand.java:64]: Removed

File: maven-scm-providers/maven-scm-providers-git/maven-scm-provider-jgit/src/main/java/org/apache/maven/scm/provider/git/jgit/command/remove/JGitRemoveCommand.java [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+import org.apache.maven.scm.command.remove.AbstractRemoveCommand;
+import org.apache.maven.scm.command.remove.RemoveScmResult;
+import org.apache.maven.scm.provider.ScmProviderRepository;
+import org.apache.maven.scm.provider.git.command.GitCommand;
+import org.apache.maven.scm.provider.git.jgit.command.JGitUtils;
+import org.eclipse.jgit.api.Git;
+
+/**
+ * @author Georg Tsakumagos
+ * @since 2.0.0-M1
```
Comment (michael-o) [inline maven-scm-providers/maven-scm-providers-git/maven-scm-provider-jgit/src/main/java/org/apache/maven/scm/provider/git/jgit/command/remove/JGitRemoveCommand.java:37]: M2

File: maven-scm-providers/maven-scm-providers-git/maven-scm-provider-jgit/src/main/java/org/apache/maven/scm/provider/git/jgit/command/JGitUtils.java [hunk 3]
```diff
@@ -461,5 +462,60 @@ public RevFilter clone()
...
+            if ( !file.isAbsolute() )
+            {
+                file = new File( fileSet.getBasedir().getPath(), file.getPath() );
+            }
+
+            if ( file.exists() )
+            {
+                String path = relativize( baseUri, file );
+                remove.addFilepattern( path );
+                remove.addFilepattern( file.getAbsolutePath() );
```
Comment (michael-o) [inline maven-scm-providers/maven-scm-providers-git/maven-scm-provider-jgit/src/main/java/org/apache/maven/scm/provider/git/jgit/command/JGitUtils.java:492]: Why is this added twice?

File: maven-scm-providers/maven-scm-providers-git/maven-scm-provider-jgit/src/main/java/org/apache/maven/scm/provider/git/jgit/command/JGitUtils.java [hunk 4]
```diff
@@ -461,5 +462,60 @@ public RevFilter clone()
...
+        Status status = git.status().call();
+
+        Set<String> allInIndex = new HashSet<String>();
+        allInIndex.addAll( status.getRemoved() );
+        List<ScmFile> removedFiles = new ArrayList<ScmFile>( allInIndex.size() );
+
+        // rewrite all detected files to now have status 'checked_in'
+        for ( String entry : allInIndex )
+        {
+            ScmFile scmfile = new ScmFile( entry, ScmFileStatus.DELETED );
```
Comment (michael-o) [inline maven-scm-providers/maven-scm-providers-git/maven-scm-provider-jgit/src/main/java/org/apache/maven/scm/provider/git/jgit/command/JGitUtils.java:506]: `scmFile`

File: maven-scm-providers/maven-scm-providers-git/maven-scm-provider-gittest/src/main/java/org/apache/maven/scm/provider/git/command/remove/RemoveCommandTckTest.java [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+    }
+    
+    public void testCommandRemoveWithFile()
+                    throws Exception
+    {
+        File workingDirectory = createTempDirectory();
+
+        File toBeRemoved = new File( workingDirectory.getAbsolutePath() + File.separator + "toto.xml" );
+        
+        Files.write(toBeRemoved.getAbsoluteFile().toPath(), "data".getBytes(StandardCharsets.UTF_8), StandardOpenOption.APPEND, StandardOpenOption.CREATE );
```
Comment (michael-o) [inline maven-scm-providers/maven-scm-providers-git/maven-scm-provider-gittest/src/main/java/org/apache/maven/scm/provider/git/command/remove/RemoveCommandTckTest.java:57]: Space around parens

File: maven-scm-providers/maven-scm-providers-git/maven-scm-provider-gittest/src/main/java/org/apache/maven/scm/provider/git/command/remove/RemoveCommandTckTest.java [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+        toBeRemoved.mkdir();
+
+        ScmFileSet fileSet =  new ScmFileSet( getWorkingCopy(), toBeRemoved );
+        RemoveScmResult removeResult = getScmManager().remove( getScmRepository(), fileSet, getBasedir() );
+        assertResultIsSuccess( removeResult );
+        
+        FileUtils.deleteDirectory( workingDirectory );
+    }
+
+    public void testCommandRemoveWithTwoDirectory()
```
Comment (michael-o) [inline maven-scm-providers/maven-scm-providers-git/maven-scm-provider-gittest/src/main/java/org/apache/maven/scm/provider/git/command/remove/RemoveCommandTckTest.java:81]: TwoDirectories

File: maven-scm-providers/maven-scm-providers-git/maven-scm-provider-gittest/src/main/java/org/apache/maven/scm/provider/git/command/remove/RemoveCommandTckTest.java [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+        ScmFileSet fileSet =  new ScmFileSet( getWorkingCopy(), Arrays.asList( toBeRemoved1, toBeRemoved2 ) );
+        RemoveScmResult removeResult = getScmManager().remove( getScmRepository(), fileSet, getBasedir() );
+        assertResultIsSuccess( removeResult );
+        FileUtils.deleteDirectory( workingDirectory );
+    }
+
+    private File createTempDirectory()
+        throws IOException
+    {
+        File dir = File.createTempFile( "gitexe", "test" );
```
Comment (michael-o) [inline maven-scm-providers/maven-scm-providers-git/maven-scm-provider-gittest/src/main/java/org/apache/maven/scm/provider/git/command/remove/RemoveCommandTckTest.java:101]: This is jgit, no?

**Review by michael-o — CHANGES_REQUESTED** [review]

**Comment by michael-o** [comment]
@G-Ork Please have a look at my comments. I want to finalize this PR.

**Comment by michael-o** [comment]
Still waiting about the duplicate addition of the file.

File: maven-scm-providers/maven-scm-providers-git/maven-scm-provider-jgit/src/main/java/org/apache/maven/scm/provider/git/jgit/command/JGitUtils.java (hunk 3 above)
Comment (michael-o) [reply maven-scm-providers/maven-scm-providers-git/maven-scm-provider-jgit/src/main/java/org/apache/maven/scm/provider/git/jgit/command/JGitUtils.java:492]: @kwin Maybe you have an explanation for this...?

Comment (kwin) [reply maven-scm-providers/maven-scm-providers-git/maven-scm-provider-jgit/src/main/java/org/apache/maven/scm/provider/git/jgit/command/JGitUtils.java:492]: RmCommand needs repository relative paths with forward slashes (https://archive.eclipse.org/jgit/site/5.0.1.201806211838-r/apidocs/org/eclipse/jgit/api/RmCommand.html#addFilepattern-java.lang.String-), therefore IMHO we need:

```
URI workingCopyRootUri = git.getRepository().getWorkTree().toURI();
String path = FilenameUtils.normalizeFilename( relativize( workingCopyRootUri, file ) );
remove.addFilepattern( path );
```

**Comment by kwin** [comment]
@michael-o I would recommend to first clarify https://github.com/apache/maven-scm/pull/152#issuecomment-1178024189 as this command also returns file paths.

**Comment by michael-o** [comment]
@kwin Please review. I have rebased and cleaned up. Also check last commit, simplification, but am not 100% sure.

File: maven-scm-providers/maven-scm-providers-git/maven-scm-provider-gittest/src/main/java/org/apache/maven/scm/provider/git/command/remove/GitRemoveCommandTckTest.java [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+    extends ScmTckTestCase
+{
+    /** {@inheritDoc} */
+    public void initRepo()
+        throws Exception
+    {
+        GitScmTestUtils.initRepo( "src/test/resources/repository/", getRepositoryRoot(), getWorkingDirectory() );
+    }
+
+    public void testCommandRemoveWithFile()
```
Comment (kwin) [inline maven-scm-providers/maven-scm-providers-git/maven-scm-provider-gittest/src/main/java/org/apache/maven/scm/provider/git/command/remove/GitRemoveCommandTckTest.java:48]: I think we use JUnit4, so a `@Test` annotation is necessary for m-surefire-p to pick this up.

File: maven-scm-providers/maven-scm-providers-git/maven-scm-provider-gittest/src/main/java/org/apache/maven/scm/provider/git/command/remove/GitRemoveCommandTckTest.java [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+        throws Exception
+    {
+        File toBeRemoved = new File( getWorkingDirectory().getAbsolutePath() + File.separator + "toto.xml" );
+
+        Files.write( toBeRemoved.getAbsoluteFile().toPath(),
```
Comment (kwin) [inline maven-scm-providers/maven-scm-providers-git/maven-scm-provider-gittest/src/main/java/org/apache/maven/scm/provider/git/command/remove/GitRemoveCommandTckTest.java:53]: why not using an empty file created with https://docs.oracle.com/javase/7/docs/api/java/nio/file/Files.html#createFile(java.nio.file.Path,%20java.nio.file.attribute.FileAttribute...), the content does not matter in this case.

File: maven-scm-providers/maven-scm-providers-git/maven-scm-provider-gittest/src/main/java/org/apache/maven/scm/provider/git/command/remove/GitRemoveCommandTckTest.java [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+            "data".getBytes( StandardCharsets.UTF_8 ), StandardOpenOption.APPEND, StandardOpenOption.CREATE );
+
+        ScmFileSet fileSet =  new ScmFileSet( getWorkingCopy(), toBeRemoved );
+        RemoveScmResult removeResult = getScmManager().remove( getScmRepository(), fileSet, getBasedir() );
+        assertResultIsSuccess( removeResult );
```
Comment (kwin) [inline maven-scm-providers/maven-scm-providers-git/maven-scm-provider-gittest/src/main/java/org/apache/maven/scm/provider/git/command/remove/GitRemoveCommandTckTest.java:58]: I think we need a consecutive checkout to be sure that `toto.xml` is really removed.

File: maven-scm-providers/maven-scm-providers-git/maven-scm-provider-jgit/src/main/java/org/apache/maven/scm/provider/git/jgit/command/JGitUtils.java [hunk 11]
```diff
@@ -527,4 +529,61 @@ public static List<String> getTags( Repository repo, RevCommit commit ) throws I
...
+    public static List<ScmFile> removeAllFiles( Git git, ScmFileSet fileSet )
+        throws GitAPIException, NoFilepatternException
+    {
+        URI baseUri = fileSet.getBasedir().toURI();
+        RmCommand remove = git.rm();
+        for ( File file : fileSet.getFileList() )
+        {
+            if ( !file.isAbsolute() )
+            {
+                file = new File( fileSet.getBasedir().getPath(), file.getPath() );
```
Comment (kwin) [inline maven-scm-providers/maven-scm-providers-git/maven-scm-provider-jgit/src/main/java/org/apache/maven/scm/provider/git/jgit/command/JGitUtils.java:552]:
```suggestion
                file = new File( fileSet.getBasedir(), file.getPath() );
```

File: maven-scm-providers/maven-scm-providers-git/maven-scm-provider-jgit/src/main/java/org/apache/maven/scm/provider/git/jgit/command/JGitUtils.java [hunk 12]
```diff
@@ -527,4 +529,61 @@ public static List<String> getTags( Repository repo, RevCommit commit ) throws I
...
+            }
+
+            if ( file.exists() )
+            {
+                URI workingCopyRootUri = git.getRepository().getWorkTree().toURI();
+                String path = FilenameUtils.normalizeFilename( relativize( workingCopyRootUri, file ) );
```
Comment (kwin) [inline maven-scm-providers/maven-scm-providers-git/maven-scm-provider-jgit/src/main/java/org/apache/maven/scm/provider/git/jgit/command/JGitUtils.java:558]: I think you figured out in the other PR that normalizeFilename is redundant in this case, as relativize always returns a path with forward slashes.

File: maven-scm-providers/maven-scm-providers-git/maven-scm-provider-jgit/src/main/java/org/apache/maven/scm/provider/git/jgit/command/JGitUtils.java [hunk 13]
```diff
@@ -527,4 +529,61 @@ public static List<String> getTags( Repository repo, RevCommit commit ) throws I
...
+        for ( String entry : allInIndex )
+        {
+            ScmFile scmFile = new ScmFile( entry, ScmFileStatus.DELETED );
+
+            // if a specific fileSet is given, we have to check if the file is
+            // really tracked
+            for ( Iterator<File> itfl = fileSet.getFileList().iterator(); itfl.hasNext(); )
+            {
+                String path = FilenameUtils.normalizeFilename( relativize( baseUri, itfl.next() ) );
+                if ( path.equals( FilenameUtils.normalizeFilename( scmFile.getPath() ) ) )
```
Comment (kwin) [inline maven-scm-providers/maven-scm-providers-git/maven-scm-provider-jgit/src/main/java/org/apache/maven/scm/provider/git/jgit/command/JGitUtils.java:580]: this logic is IMHO incorrect, as `scmFile` has repository relative path names (already with forward slashes) while `path` is relative to basedir.

File: maven-scm-providers/maven-scm-providers-git/maven-scm-provider-gittest/src/main/java/org/apache/maven/scm/provider/git/command/remove/GitRemoveCommandTckTest.java [hunk 14]
```diff
@@ -0,0 +1,85 @@
...
+        File toBeRemoved1 = new File( getWorkingDirectory().getAbsolutePath() + File.separator + "toto" );
+        toBeRemoved1.mkdir();
+
+        File toBeRemoved2 = new File( getWorkingDirectory().getAbsolutePath() + File.separator + "tata" );
+        toBeRemoved2.mkdir();
+
+        ScmFileSet fileSet =  new ScmFileSet( getWorkingCopy(), Arrays.asList( toBeRemoved1, toBeRemoved2 ) );
+        RemoveScmResult removeResult = getScmManager().remove( getScmRepository(), fileSet, getBasedir() );
+        assertResultIsSuccess( removeResult );
+    }
```
Comment (kwin) [inline maven-scm-providers/maven-scm-providers-git/maven-scm-provider-gittest/src/main/java/org/apache/maven/scm/provider/git/command/remove/GitRemoveCommandTckTest.java:84]: we should add a test for a basedir != repository root

**Review by kwin — CHANGES_REQUESTED** [review]

File: maven-scm-providers/maven-scm-providers-git/maven-scm-provider-gittest/src/main/java/org/apache/maven/scm/provider/git/command/remove/GitRemoveCommandTckTest.java (hunk 9 above)
Comment (michael-o) [reply maven-scm-providers/maven-scm-providers-git/maven-scm-provider-gittest/src/main/java/org/apache/maven/scm/provider/git/command/remove/GitRemoveCommandTckTest.java:53]: Does it matter, because Git tracks content, not files?!

File: maven-scm-providers/maven-scm-providers-git/maven-scm-provider-gittest/src/main/java/org/apache/maven/scm/provider/git/command/remove/GitRemoveCommandTckTest.java [hunk 15]
```diff
@@ -0,0 +1,85 @@
...
+    public void initRepo()
+        throws Exception
+    {
+        GitScmTestUtils.initRepo( "src/test/resources/repository/", getRepositoryRoot(), getWorkingDirectory() );
+    }
+
+    public void testCommandRemoveWithFile()
+        throws Exception
+    {
+        File toBeRemoved = new File( getWorkingDirectory().getAbsolutePath() + File.separator + "toto.xml" );
```
Comment (kwin) [inline maven-scm-providers/maven-scm-providers-git/maven-scm-provider-gittest/src/main/java/org/apache/maven/scm/provider/git/command/remove/GitRemoveCommandTckTest.java:51]:
```suggestion
        File toBeRemoved = new File( getWorkingDirectory(), "toto.xml" );
```

File: maven-scm-providers/maven-scm-providers-git/maven-scm-provider-gittest/src/main/java/org/apache/maven/scm/provider/git/command/remove/GitRemoveCommandTckTest.java (hunk 9 above)
Comment (kwin) [reply maven-scm-providers/maven-scm-providers-git/maven-scm-provider-gittest/src/main/java/org/apache/maven/scm/provider/git/command/remove/GitRemoveCommandTckTest.java:53]: Right, then maybe simplify to 
```suggestion
        Files.write( toBeRemoved.toPath(), Collections.singletonList("data"), StandardCharsets.ASCII, StandardOpenOption.CREATE )
```

**Comment by michael-o** [comment]
I think the TCK test is logically flawed. It removes files and dirs:
* File has never been added to SCM
* You cannot remove dirs with Git if there is no content. Not visible to Git
* It is a copy of `org.apache.maven.scm.provider.git.gitexe.command.remove.GitRemoveCommandTest` which checks the generated CLI.

**Comment by michael-o** [comment]
@kwin Applied most of your comments
´

File: maven-scm-providers/maven-scm-providers-git/maven-scm-provider-jgit/src/main/java/org/apache/maven/scm/provider/git/jgit/command/JGitUtils.java [hunk 16]
```diff
@@ -576,8 +576,8 @@ public static List<ScmFile> removeAllFiles( Git git, ScmFileSet fileSet )
             // really tracked
             for ( Iterator<File> itfl = fileSet.getFileList().iterator(); itfl.hasNext(); )
             {
-                String path = FilenameUtils.normalizeFilename( relativize( baseUri, itfl.next() ) );
-                if ( path.equals( FilenameUtils.normalizeFilename( scmFile.getPath() ) ) )
+                String path = relativize( baseUri, itfl.next() );
+                if ( path.equals( scmFile.getPath() ) )
```
Comment (kwin) [inline maven-scm-providers/maven-scm-providers-git/maven-scm-provider-jgit/src/main/java/org/apache/maven/scm/provider/git/jgit/command/JGitUtils.java:580]: this comparison is still wrong, IMHO you first need to make `itfl.next()` absolute and then relativize against `workingCopyRootUri`.

Comment (michael-o) [reply maven-scm-providers/maven-scm-providers-git/maven-scm-provider-jgit/src/main/java/org/apache/maven/scm/provider/git/jgit/command/JGitUtils.java:580]: Correct. Please fork and provide a PR on top. You know better what to do already.

**Comment by michael-o** [comment]
The gitexe test is incomplete as well.

**Comment by michael-o** [comment]
@kwin Should we stall this for M2 until we get a clearer picture and @G-Ork can chime in?

**Comment by kwin** [comment]
I may find some time next week to look into that. If you want to release M2 earlier just defer this.

**Comment by michael-o** [comment]
> I may find some time next week to look into that. If you want to release M2 earlier just defer this.

Thank you. I will defer and leave you and @G-Ork enough time to crunch on this. Especially because we have `gitexe` with works as an alternative. I will continue with the release today.

**Comment by kwin** [comment]
@michael-o You find my refinement in #156.

**Comment by jira-importer** [comment]
Resolve #1150
